File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()

        closest = nums[0] + nums[1] + nums[2]

        for k in range(len(nums) - 2):
            if k > 0 and nums[k] == nums[k - 1]:
                continue

            i = k + 1
            j = len(nums) - 1

            while i < j:
                curr_sum = nums[k] + nums[i] + nums[j]

                if abs(curr_sum - target) < abs(closest - target):
                    closest = curr_sum

                if curr_sum == target:
                    return target
                elif curr_sum > target:
                    j -= 1
                else:
                    i += 1

        return closest

File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {

        Arrays.sort(nums);

        int closest = nums[0] + nums[1] + nums[2];

        for (int k = 0; k < nums.length - 2; k++) {

            if (k > 0 && nums[k] == nums[k - 1])
                continue;

            int i = k + 1;
            int j = nums.length - 1;

            while (i < j) {

                int sum = nums[k] + nums[i] + nums[j];

                if (Math.abs(sum - target) < Math.abs(closest - target)) {
                    closest = sum;
                }

                if (sum == target) {
                    return target;
                } else if (sum > target) {
                    j--;
                } else {
                    i++;
                }
            }
        }

        return closest;
    }
}